### ISSUE TITLE
Softplus

### ISSUE THREAD

**Issue by edmondja** [brownvc/R3GAN#10]
Are we sure of 
RelativisticLogits = FakeLogits - RealLogits 
AdversarialLoss = nn.functional.softplus(-RelativisticLogits)
instead of minimizing
AdversarialLoss = -nn.functional.softplus(-RelativisticLogits)
for the generator ?
Because the paper says "G tries to minimize L"
and 
L = − log(1 + exp(−t)) = - log(1 + exp(−RelativisticLogits)) = -nn.functional.softplus(-RelativisticLogits)

Also the discriminator loss "L while D tries to maximize it"
so the loss we need to minimize must be 
RelativisticLogits = FakeLogits - RealLogits 
AdversarialLoss = nn.functional.softplus(-RelativisticLogits) don't you think ?

**Comment by nick-bvc** [comment]
The implementation is correct. We follow the notation of Mescheder et al. in the paper, there might be a typo in the original paper: https://arxiv.org/abs/1801.04406

**Comment by edmondja** [comment]
> The implementation is correct. We follow the notation of Mescheder et al. in the paper, there might be a typo in the original paper: https://arxiv.org/abs/1801.04406

In the paper you just quoted and in the paper "The GAN is dead; long live the GAN!" we suggest to make the discriminator maximize the function the generator is minimizing. 
This is not what the code is doing, as minimizing f(x) and maximizing -f(x) are equivalent, but minimizing f(x) and maximizing f(-x) (what is done in the code) are not. Please confirm me you are aware of this and that it was actually done on purpose in the code and that I am the one being wrong here.

**Comment by nick-bvc** [comment]
the code is correct. The analyses of GANs (including our paper, Mescheder et al., Sun et al.) are typically on saturating GANs, i.e., -f(x), but non-saturating GANs, i.e., f(-x) are what's used in practice.
Non-saturating GANs have better gradient behavior and regularization effects, in fact, even unregularized non-saturating GANs can converge at a logarithmic rate in scenarios where unregularized saturating GANs fail.

**Comment by edmondja** [comment]
Wow I wouldnt guess from reading the paper, very interesting thank you so much for your time and for this amazing paper
